Support Half type in randperm.

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22121 Refactor and improve randperm tests.
* #22103 Swap detection order in randperm_out_cuda to avoid unnecessary conversion from float when the input is small.
* **#22102 Support Half type in randperm.**


Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

Differential Revision: [D16153586](https://our.internmc.facebook.com/intern/diff/D16153586)

xuhdev committed:
Support Half type in randperm.

xuhdev commented:
@pytorchbot retest this please

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev commented:
@pytorchbot retest this please

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

gchanan commented on aten/src/ATen/native/TensorFactories.h:
```diff
@@ -64,5 +64,25 @@ inline void check_size_nonnegative(IntArrayRef size) {
+
+  // Ensure sufficient precision for floating point representation.
+  switch (tensor.scalar_type()) {
+  case at::ScalarType::Half:
```
nit: indent

xuhdev replied:
Fixed

gchanan commented on test/test_torch.py:
```diff
@@ -4427,11 +4427,19 @@ def test_randperm(self):
         # randperm of 0 elements is an empty tensor
         res1 = torch.randperm(0)
-        res2 = torch.LongTensor(5)
+        res2 = torch.LongTensor()
```
why did this change?

gchanan commented on aten/src/ATen/native/TensorFactories.h:
```diff
@@ -64,5 +64,25 @@ inline void check_size_nonnegative(IntArrayRef size) {
 }
+
+inline void check_supported_max_int_with_precision(int64_t n, const Tensor& tensor) {
+  TORCH_CHECK(at::scalar_tensor(n, tensor.options()).defined(),
```
I'm confused about the error checking here.  Do you know if it's actually possible to trigger this TORCH_CHECK?  I can't see how it is, but maybe I'm missing something.

xuhdev replied:
This was moved out from the original CUDA implementation. To me it seems checking for situation like the following:

    torch.randperm(100000, dtype=torch.half, device='cuda')                                                               

(half cannot represent a number as large as 100000.)

Practically it gives me a different error (before I made this change, so it would check for the cuda function only):
                               
    RuntimeError: value cannot be converted to type at::Half without overflow: 100000

It doesn't seem helpful, but IMO it won't hurt to stay either.

gchanan replied:
sure, sounds reasonable.

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

gchanan commented:
CI failures look unrelated.

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head

xuhdev committed:
Update on "Support Half type in randperm."

Support Half type in randperm.

Previously randperm supports Half type only on CUDA. This commit adds
Half support for the CPU version. Precision check for floating point
type is also added to ensure that integers can be properly represented.

gh-metadata: pytorch pytorch 22102 gh/xuhdev/1/head
